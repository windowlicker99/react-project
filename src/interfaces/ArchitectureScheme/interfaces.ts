import { NodeModel, ISelectionChangeEventArgs } from '@syncfusion/ej2-diagrams';
import { ICurrentElement, IDomain, IECUBasicFields } from '../interfaces';
import { TDiagramEventElement } from './types';

export interface IArchitectureSchemeDomain {
  id: string;
  name: string;
  color: string;
  ECUs?: IArchitectureSchemeECU[];
}

export interface IArchitectureSchemeECU {
  id: string;
  title: string;
  color: string;
  isGateway?: boolean;
  connectedECUs?: IArchitectureSchemeECU[];
  relativeECUs?: IArchitectureSchemeECU[];
}

export interface IArchitectureSchemeECUasGateway extends IArchitectureECU {
  domainNodes: NodeModel[];
}

export interface IArchitecturePointsRange {
  min: number;
  max: number;
}

export interface IDiagramTooltip {
  background?: string;
  color?: string;
  text: string;
}

export interface IDiagramRelative {
  color: string;
}

export interface IElementAddInfo {
  element: ICurrentElement;
  tooltip?: IDiagramTooltip;
  relative?: IDiagramRelative;
}

export interface IDiagramClickEventArgs {
  element: TDiagramEventElement;
  actualObject?: TDiagramEventElement;
}

export interface ISelectionChangeArgs extends ISelectionChangeEventArgs {
  newValue: TDiagramEventElement[];
  oldValue: TDiagramEventElement[];
}

export interface IArchitectureECU extends IECUBasicFields {
  id: string;
  vehicleId: string;
  ticketId?: string | number;
  connectedECUs: IArchitectureECU[];
  connectedVMs?: IArchitectureECU[];
  relativeECUs?: IArchitectureECU[];
  color?: string;
  domain?: { id: string };
}
export interface IArchitectureDomain extends Partial<IDomain> {
  ECUs: IArchitectureECU[];
}
