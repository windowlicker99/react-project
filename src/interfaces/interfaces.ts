import { ReactElement, ReactNode } from 'react';
import { Column } from 'react-table';
import {
  EApproved,
  EAside,
  EController,
  EDomain,
  EPhase,
  ERegulations,
  EDirectCommunication,
  EECUTypes,
  EImplemented,
  ERiskLevels,
  EUpdateProcessRadio,
  ETicketsTypes,
  EExpertise,
  EAccess,
  EEquipment,
  EAttackDifficultyFinalScope,
  EPhaseStatus,
  EGatewayTypes,
} from './enums';
import { TOnlineConnectivity, TList } from './types';

export enum ETableTypes {
  domains = 'domains',
  ECUs = 'ecus',
  functions = 'functions',
  data = 'data',
  vehicles = 'vehicles',
  conceptThreats = 'conceptThreats',
  threats = 'threats',
  mainFindings = 'mainFindings',
  detailedFindings = 'detailedFindings',
  // conceptThreats = 'threats',
  securityConcepts = 'securityConcepts',
  controlGroups = 'controlGroups',
  feeds = 'feeds',
  threatsOem = 'threatsOem',
  controlGroupsOem = 'controlGroupsOem',
}

export interface IBrand {
  id?: string;
  name: string;
  fullName: string;
  logo: string;
}

export interface IGateway {
  id: string;
  name: string;
  safetyCriticalityLevel: string;
  firmwareVersion: string;
  onlineConnectivity: TOnlineConnectivity;
  operationSystem: string;
  vehicleId?: string;
}

export interface IVehicleBasicFields {
  model: string;
  platform: string;
}

export interface IVehicle extends IVehicleBasicFields {
  id: string;
  manufacturer: string;
  name: string;
  riskLevel: keyof typeof ERiskLevels;
  raProgress: number;
  brand: IBrand;
  complianceLevel: number;
  lastUpdate: string;
  logo: string;
  photo: string;
  phase: EPhase;
  regulation: ERegulations;
  development: string;
  testing: string;
  compliancy: string;
  production: string;
}

export interface IVehicleFormValues {
  photo: File;
  brand: IOption;
  model: string;
  platform: string;
  phase: IOption;
  regulation: IOption;
  haveGateway: boolean;
  gateway?: IGatewayFormValues;
  development: Date;
  compliancy: Date;
  testing: Date;
  production: Date;
}

export interface IGatewayFormValues {
  name: string;
  safetyCriticalityLevel: IOption;
  operationSystem: string;
  firmwareVersion: string;
  connectivity: boolean;
}

export interface IPostedVehicle extends IVehicleBasicFields {
  brandId: string;
  photo: File;
  gateway?: Partial<IGateway>;
  development: Date;
  testing: Date;
  compliancy: Date;
  production: Date;
  phase: string;
  regulation: string;
  haveGateway: boolean;
}

export interface IAuthor {
  id: string;
  name: string;
  password: string;
  avatar: string;
}

export interface IPostDomain {
  vehicleId: string;
  name: string;
  domainUsage: string;
  communication: string[];
  onlineConnectivity: TOnlineConnectivity;
  color: string;
  background: string;
  authorId: string;
}

export interface IGatewayResponse {
  type: EGatewayTypes;
  gateway: IGateway | IECUasGateway;
}

export interface IDomain {
  id: string;
  model: string;
  platform: string;
  name: string;
  domainUsage: string;
  communication: string[];
  onlineConnectivity: TOnlineConnectivity;
  riskLevel?: ERiskLevels;
  author: IAuthor;
  color: string;
  background: string;
  lastUpdate: string;
}

export interface ISupplier {
  id: string;
  name: string;
}

export interface ICommunication {
  id: string;
  communication: string;
}

export interface IECUBasicFields {
  name: string;
  complianceLevel?: number;
  operationSystem?: string;
  firmwareVersion?: string;
  onlineConnectivity: TOnlineConnectivity;
  hostVMorDocker: string | boolean;
  runVMorDocker: string | boolean;
  haveDirectCommunication?: boolean;
  updateProcess?: boolean;
  managingOrUsing?: EUpdateProcessRadio;
  storeConfidentialData?: boolean;
  useCryptographicTechnologies?: boolean;
  indirectCommunicationChannel?: boolean;
  hostECUId?: string;
  isGateway?: boolean;
  isSwitch?: boolean;
}

export interface IECUBasicInfoField {
  type: EECUTypes;
  directCommunication?: EDirectCommunication[];
}

export interface IECUFormValues extends IECUBasicFields {
  supplierId: string;
  hostECU: IOption;
  isRegular: boolean;
  supplier: IOption;
  domains: IOptionsList[];
  directCommunication: IOption[];
  connectedECUs: IOptionsList[];
}

export interface IPostedECU extends IECUBasicFields, IECUBasicInfoField {
  supplierId: string;
  domains?: ICommunication[];
  connectedECUs?: ICommunication[];
}

export interface ITicket {
  name: string;
  userId: string;
  vehicleId: string;
  elementType: string;
  elementName: string;
  elementId?: string;
  message?: string;
  user?: IAuthor;
  vehicle?: IVehicle;
  status: string;
  type: ETicketsTypes;
  created: string;
  updated: string;
}

export interface IECU extends IECUBasicFields, IECUBasicInfoField {
  id: string;
  vehicle: IVehicle;
  domains: IDomain[];
  supplier: ISupplier;
  complianceLevel: number;
  ticket: ITicket;
  safetyCriticalityLevel: string;
  riskLevel: ERiskLevels;
  connectedECUs: IECU[];
  relativeECUs?: IECU[];
  connectedVMs?: IECU[];
  vehicleId: string;
}

export interface IECUasGateway extends IECU {
  color?: string;
  background?: string;
}

export interface IFunctions {
  id: string;
  model: string;
  platform: string;
  domains: IDomain[];
  ecu: IECU;
  name: string;
  supplier: string;
  complianceLevel: number;
  version: string;
  cluster: string;
  connectivity: string;
}

export interface IData {
  id: string;
  model: string;
  platform: string;
  domains: IDomain[];
  ecu: IECU;
  functions: IFunctions;
  name: string;
}

export interface IPropertiesData {
  [key: string]: string | number;
}

export interface IColumn {
  Header: string;
  accessor: string;
  name?: string;
}

export interface IFilter {
  name: string;
  label: string;
}

export interface IUser {
  id: string;
  name: string;
  avatar: string;
}

export interface ILoginData {
  name: string;
  password: string;
}

export interface IElementsTab {
  name: string;
  columns: Column[];
  type: ETableTypes;
}

export interface IElementsTabsColumnsList {
  [key: string]: IColumn[];
}

export interface IVehicleRow {
  brand: string;
  model: string;
  platform: string;
  phase: EPhase;
  riskLevel: ReactElement;
  complianceLevel: ReactElement;
  raProgress: ReactElement;
  lastUpdate: ReactElement;
  actions: ReactElement;
  logo: string;
  photo: string;
  name: string;
}

export interface IDomainRow {
  model: string;
  platform: string;
  name: string;
  usage: string;
  communication: string;
  connectivity: string;
  author: ReactElement;
  update: ReactElement;
  color: ReactElement;
  actions: ReactElement;
}

export interface IECURow {
  name: string;
  type: string;
  model: string;
  platform: string;
  domain: string;
  supplier: string;
  complianceLevel: ReactElement;
  operationSystem: string;
  firmwareVersion: string;
  connectivity: string;
  actions: ReactElement;
}

export interface IFunctionsRow {
  model: string;
  platform: string;
  domain: string;
  ecu: string;
  name: string;
  supplier: string;
  level: ReactElement;
  version: string;
  cluster: string;
  connectivity: string;
  actions: ReactElement;
}

export interface IDataRow {
  model: string;
  platform: string;
  domain: string;
  ecu: string;
  function: string;
  name: string;
}

export interface IThreatRow {
  name: string;
  category: string;
  calculatedAttackFeasibility: string | number;
  classificationGroup: string;
  securityObjective: string;
  subRows?: { control: IControl }[];
}

export interface IThreatListRow {
  name: ReactElement;
  actions: ReactElement;
}

export interface IVehicleFeed {
  brand: IBrand;
  model: string;
  platform: string;
  riskLevel: keyof typeof ERiskLevels;
  ECUs: IECU;
  riskLevelProgress: IRiskLevelProgressData[];
}

export interface IFeedRow {
  brand: string;
  model: string;
  platform: string;
  riskLevel: ReactElement;
  status: ReactElement;
  ECUs: IECU;
}

export interface IMainFindingRow {
  id: string;
  topic: string;
  relatedRisk: string;
  threat: string;
  recommendations: string;
  attackFeasibility: string;
  damagePotential: string;
  comments: string;
  ticketNumber: string;
}

export interface IDetailedFindingRow {
  id: string;
  topic: string;
  status: string;
  type: string;
  purpose: string;
  description: string;
  threat: string;
  recommendations: string;
  expertise: string;
  access: string;
  time: string;
  equipmentRepetition: string;
  targetRepetition: string;
  sumRepetition: string;
  attackDifficultyFinalScope: string;
  cve: string;
}

export interface IConceptRow {
  name: string;
  description: ReactElement | string;
  subRows: { subRequirement: IRequirement }[];
}

export interface IControlsRow {
  name: string;
  actions: ReactElement;
  subRows: { control: IControl }[];
}

export interface IControlGroups {
  name: string;
  controls: IControl[];
}

export interface IOption<T = string> {
  label: T;
  value: T;
  isDisabled?: boolean;
}

export interface IField {
  label: string;
  formLabel?: string;
  placeholder?: string;
  defaultValue?: TList<Partial<IOption>>;
  initialValue?: string | boolean | TList<Partial<IOption>>[];
  values?: string[];
  name: string;
  component?: ReactNode;
  options?: IOption[];
  isMulti?: boolean;
  isSelect?: boolean;
  disabled?: boolean;
  maxLength?: number;
  subFields?: {
    [key: string]: IField;
  };
}

export interface IDateField extends IField {
  minDate?: string;
}

export interface IFieldsList {
  [key: string]: IField;
}

export interface ICurrentElement {
  type: EAside;
  id: string;
}

export interface IProperties {
  name: string;
  value?: string;
}

export interface IControllerProperties extends IProperties {
  accessor: EController;
}

export interface IDomainProperties extends IProperties {
  name: string;
  value?: string;
  accessor: EDomain;
}

export interface IOptionsList {
  [key: string]: IOption;
}

export interface ISecurityConcept {
  id: string;
  name: string;
  descriptions?: string[];
  subRequirements?: IRequirement[];
}

export interface IRequirement {
  id: string;
  reqId: string;
  name?: string;
  type?: string;
  description?: string;
}

export interface IPenetrationTest {
  id: string;
  name: string;
}

export interface IAlert {
  id: string;
  name: string;
  planned: Date;
  started: Date;
  url: string;
  vehicleIds: string[];
}

export interface IParams {
  [key: string]: string;
}

export interface ISecurityDevelopmentTab {
  name: string;
  params: IParams;
  type: ETableTypes;
}

export interface IControl {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  id: number;
  name: string;
  category: string;
  subCategory?: string;
  checked: boolean;
  comment: string;
  details: IControlsDetails;
}

export enum EToggleTreatsTabTypes {
  UNECE = 'UNECE',
  OEM = 'OEM',
}

export interface IToggleTab {
  name: string;
  columns: Column[];
  type: EToggleTreatsTabTypes;
}

export interface IThreatsDetails {
  group: string;
  calculatedAttackFeasibility: string | number;
  classificationGroup: string;
  securityObjective: string;
  technologies: string;
  dataECUThreat: string;
  elapsedTime: string;
  expertise: string;
  knowledge: string;
  equipment: string;
  windowsOfOpportunityAccess: string;
  affectLevel?: string;
}

export interface IThreat {
  id: number;
  name: string;
  category: string;
  checked: true;
  comment: string;
  details: IThreatsDetails;
  controls: IControl[];
}

export interface IControlsDetails {
  relatedThreats: string;
  calculatedControlStrength: string;
  mitigationGroups: string[];
  affectLevel: string;
  elapsedTime: string;
  expertise: string;
  knowledge: string;
  equipment: string;
  windowsOfOpportunity: string;
}

export interface IHeader {
  expandedDepth: number;
  toggleAllRowsExpanded: (isExpanded: boolean) => void;
}

export interface ISecurityConceptControlValues {
  approved: EApproved;
  implemented: EImplemented;
  comment: string;
}

export interface IMainFinding {
  id: string;
  topic: string;
  relatedRisk: string;
  threat: string;
  recommendations: string;
  attackFeasibility: string;
  damagePotential: string;
  comments: string;
  ticketNumber: string;
}

export interface IDetailedFinding {
  id: string;
  topic: string;
  status: string;
  type: string;
  purpose: string;
  description: string;
  threat: string;
  recommendations: string;
  expertise: EExpertise;
  access: EAccess;
  time: string;
  equipmentRepetition: EEquipment;
  targetRepetition: string;
  sumRepetition: string;
  attackDifficultyFinalScope: EAttackDifficultyFinalScope;
  cve: string;
}

export interface ISubRequirementsValues {
  approved: EApproved;
  implemented: EImplemented;
  comment: string;
  supplierComment: string;
}

export interface IRiskLevelProgressData {
  id: string;
  riskLevel: keyof typeof ERiskLevels;
  date: string;
}
export interface IChartParams {
  height: number;
  innerRadius: number;
  outerRadius: number;
  startAngle: number;
  value: number;
  text: string;
  className?: string;
}

export interface IProgressPoint {
  title: string;
  startDate?: string;
  date?: string;
  status?: EPhaseStatus;
}

export interface INamedObject {
  name: string;
}

export interface IIRAComment {
  text: string;
  authorId: string;
  date: string;
  authorName?: string;
}

export interface ICommentWithAuthor {
  comment: IIRAComment;
  checked: boolean;
  authorName: string;
}

export interface IComment {
  text: string;
  authorId: string;
  date: string;
}
