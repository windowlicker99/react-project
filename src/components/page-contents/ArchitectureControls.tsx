import React, { FC, useEffect, useState } from 'react';
import { Btn } from '@components/common/form-controls/Btn';
import { Badge } from '@components/common/Badge';
import { Legend } from '@components/common/Legend';
import { PageFilter } from '@components/page-contents/PageFilter';
import { CircleChart } from '@components/common/CircleChart';
import { domainFilters } from '@constants/Filters';
import { IVehicle } from '@interfaces/interfaces';
import { EBadgeTypes, ERiskLevels } from '@interfaces/enums';
import { initialVehicle } from '@constants/initialData';

interface IArchitectureControls {
  vehicle: IVehicle;
  onOpenDomainModalClick: () => void;
  onOpenHistoryView: () => void;
}

export const ArchitectureControls: FC<IArchitectureControls> = ({
  vehicle,
  onOpenDomainModalClick,
  onOpenHistoryView,
}) => {
  const [displayedVehicle, setDisplayedVehicle] = useState<IVehicle>(initialVehicle);
  const [isOpenFilters, setIsOpenFilters] = useState<boolean>(false);

  const onShowFiltersClick = () => {
    setIsOpenFilters(!isOpenFilters);
  };

  const refreshFilters = () => {
    // temporary solution. Will be removed with adding filters refresh logic
    console.log('refresh filters');
  };

  useEffect(() => {
    if (!vehicle) {
      setDisplayedVehicle(initialVehicle);
      return;
    }
    setDisplayedVehicle(vehicle);
  }, [vehicle]);

  return (
    <div className="box page-controls">
      <div className="page-controls-content">
        <div className="page-controls-item">
          <div className="mr-15">
            <Btn
              disabledBtn={!vehicle}
              btnStatus="btn-lg-primary"
              btnIcon="icon-plus"
              btnText="Create Domain"
              onClick={onOpenDomainModalClick}
            />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-lg-outline" btnText="View History" onClick={onOpenHistoryView} />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-xs-primary" btnIcon="icon-filter-blue" onClick={onShowFiltersClick} />
          </div>
          <div className="mr-15">
            <Btn btnStatus="btn-xs-secondary" btnIcon="icon-refresh" onClick={refreshFilters} />
          </div>
        </div>
        <div className="page-controls-item">
          <div className="page-controls-architecture">
            <div className="page-controls-data">
              {vehicle && (
                <div className="page-controls-info">
                  <span className="page-controls-name">{displayedVehicle.brand?.name}</span>
                  {displayedVehicle.model}
                </div>
              )}
            </div>
            <div className="page-controls-data">
              Risk Level
              <Badge
                badgeType={EBadgeTypes[displayedVehicle.riskLevel]}
                badgeText={ERiskLevels[displayedVehicle.riskLevel]}
              />
            </div>
            <div className="page-controls-data">
              Progress
              <div className="chart">
                <CircleChart percentage={displayedVehicle.raProgress} />
              </div>
            </div>
            <div className="page-controls-data">
              <Legend isDisabled={!vehicle} legendType="Eth" legendName="Ethernet" />
              <Legend isDisabled={!vehicle} legendType="CAN" legendName="CAN" />
              <Legend isDisabled={!vehicle} legendType="LIN" legendName="LIN" />
            </div>
          </div>
        </div>
      </div>
      <PageFilter filters={domainFilters} isOpen={isOpenFilters} close={onShowFiltersClick} />
    </div>
  );
};
