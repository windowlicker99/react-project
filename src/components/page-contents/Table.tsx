import React, { FC, useEffect, useState } from 'react';
import { useTable, CellValue, Column, useExpanded, usePagination } from 'react-table';
import { useDispatch } from 'react-redux';
import { ControlDropdown } from '@components/common/ControlDropdown';
import { TableDropdown } from '@components/common/TableDropdown';
import { RequirementDropdown } from '@components/common/RequirementDropdown';
import { PropertiesCard } from '@components/common/PropertiesCard';
import { ControlListDropdown } from '@components/common/ControlListDropdown';
import { FeedsECUDropdown } from '@components/common/FeedsECUDropdown';
import {
  setNextPage,
  setPrevPage,
  currentPage,
  pageAmount,
  pickPage,
  disableNext,
  disablePrev,
  currentTab,
  totalElement,
  totalElementOnPage,
} from '@store/slices/paginationSlice';
import { changeType } from '@store/slices/columnsSlice';
import { getTableItems, getSearchResults } from '@utils/getTableItems';
import { getColumnsWithFilter } from '@utils/getColumnsWithFilter';
import { ETableTypes, IParams } from '@interfaces/interfaces';
import { TRows } from '@interfaces/types';
import { EThreatsControlsPath } from '@interfaces/enums';

interface ITableProps {
  tableType: ETableTypes;
  changeLoadingState: (props: boolean) => void;
  columns: Column[];
  searchText?: string;
  setSearchText?: (arg0: string) => void;
  updateContent?: boolean;
  params?: IParams;
  showHeader?: boolean;
  withoutPagination?: boolean;
}

export const Table: FC<ITableProps> = ({
  updateContent,
  changeLoadingState,
  tableType,
  params = {},
  columns,
  searchText,
  setSearchText,
  showHeader = false,
  withoutPagination = false,
}) => {
  const dispatch = useDispatch();
  const [tableRows, setTableRows] = useState<TRows>([]);
  const [previousExpandedRows, setPreviousExpandedRows] = useState([]);
  const [columnsWithFilter, setColumnsWithFilter] = useState([]);

  const getRows = async () => {
    changeLoadingState(true);

    const items = await getTableItems(tableType, params);
    setTableRows(items);
    changeLoadingState(false);
  };

  const getSearchedRows = async () => {
    changeLoadingState(true);

    const items = await getSearchResults(tableType, searchText);
    setTableRows(items);

    changeLoadingState(false);
  };

  useEffect(() => {
    if (searchText) {
      getSearchedRows();
      return;
    }
    getRows();
  }, [searchText]);

  useEffect(() => {
    getRows();
    dispatch(changeType({ type: tableType }));
    const columnsNames = getColumnsWithFilter(tableType);
    setColumnsWithFilter(columnsNames);

    if (searchText !== undefined) {
      setSearchText('');
    }
  }, [tableType, updateContent]);

  const {
    rows,
    getTableProps,
    getTableBodyProps,
    headerGroups,
    page,
    canPreviousPage,
    canNextPage,
    nextPage,
    previousPage,
    gotoPage,
    prepareRow,
    setHiddenColumns,
    state: { pageIndex },
    setPageSize,
  } = useTable(
    {
      columns,
      data: tableRows,
      initialState: {
        pageIndex: 0,
        pageSize: 8,
      },
    },
    useExpanded,
    usePagination
  );

  useEffect(() => {
    if (withoutPagination && rows?.length) {
      setPageSize(rows.length);
    }
  }, [rows]);

  const collapseSubRows = () =>
    previousExpandedRows.forEach((row) => {
      const idParts = row.id.split('.');

      if (idParts.length < 2) {
        return;
      }

      const parent = rows.find((row) => row.id === idParts[0]);

      if (parent && !parent.isExpanded && row.isExpanded) {
        row.toggleRowExpanded();
      }
    });

  const capitalizeFirstLetter = (string: string) => string.charAt(0).toUpperCase() + string.slice(1);

  useEffect(() => {
    if (
      tableType === ETableTypes.conceptThreats ||
      tableType === ETableTypes.securityConcepts ||
      tableType === ETableTypes.controlGroups
    ) {
      collapseSubRows();
      setPreviousExpandedRows(rows);
    }
  }, [rows]);

  dispatch(setNextPage(nextPage));
  dispatch(setPrevPage(previousPage));
  dispatch(pickPage(gotoPage));
  dispatch(currentPage(pageIndex));
  dispatch(pageAmount(Math.ceil(tableRows.length / 8)));
  dispatch(disablePrev(canPreviousPage));
  dispatch(disableNext(canNextPage));
  dispatch(
    currentTab(
      (tableType === 'controlGroups' && 'Controls') ||
        (tableType === 'threatsOem' && 'Threats') ||
        (tableType === 'controlGroupsOem' && 'Controls') ||
        capitalizeFirstLetter(tableType)
    )
  );
  dispatch(totalElement(tableRows.length));
  dispatch(totalElementOnPage(page.length));

  useEffect(() => {
    let columnsToHide: string[] = [];
    if (tableType === ETableTypes.vehicles) {
      columnsToHide = ['name', 'id'];
    }
    setHiddenColumns([
      'photo',
      'logo',
      'control',
      'controlInfo',
      'details',
      'ECUs',
      'development',
      'production',
      'testing',
      'compliancy',
      ...columnsToHide,
    ]);
  }, [columns]);

  return (
    <>
      <table className="table" {...getTableProps()}>
        {showHeader && (
          <thead className="table-header">
            {headerGroups.map((headerGroup) => (
              <tr key={headerGroup.id} {...headerGroup.getHeaderGroupProps()} className="table-row">
                {headerGroup.headers.map((column) => (
                  <th className="table-col" {...column.getHeaderProps()} key={column.id}>
                    <div className="table-header-inner">
                      <span className="table-header-title">{column.render('Header')}</span>
                      <div className="table-filter">
                        {columnsWithFilter.includes(column.id) && (
                          <>
                            <button type="button" className="table-filter-btn">
                              <span className="icon-arrow-full-up" />
                            </button>
                            <button type="button" className="table-filter-btn">
                              <span className="icon-arrow-full-down" />
                            </button>
                          </>
                        )}
                      </div>
                    </div>
                  </th>
                ))}
              </tr>
            ))}
          </thead>
        )}
        <tbody className="table-body" {...getTableBodyProps()}>
          {page.map((row) => {
            prepareRow(row);

            switch (tableType) {
              case ETableTypes.vehicles:
                return (
                  <React.Fragment key={row.id}>
                    <tr {...row.getRowProps()} className="table-row">
                      {row.cells.map((cell: CellValue) => (
                        <td className="table-col" {...cell.getCellProps()} key={cell.column.id}>
                          {cell.render('Cell')}
                        </td>
                      ))}
                    </tr>
                    {row.isExpanded && (
                      <tr>
                        <td colSpan={99}>
                          <TableDropdown
                            id={row.values.id}
                            name={row.values.name}
                            logo={row.values.logo}
                            photo={row.values.photo}
                            model={row.values.model}
                            currentPhase={row.values.phase}
                            percentage={row.values.complianceLevel.props.percentage}
                          />
                        </td>
                      </tr>
                    )}
                  </React.Fragment>
                );
              case ETableTypes.conceptThreats:
                return (
                  <React.Fragment key={row.id}>
                    {row.values.control ? (
                      <ControlDropdown
                        control={row.values.control}
                        showExpandedInfo={row.isExpanded}
                        expander={row.cells[0]}
                      />
                    ) : (
                      <tr {...row.getRowProps()} className="table-row" key={row.id}>
                        {row.cells.map((cell: CellValue) => (
                          <td className="table-col" {...cell.getCellProps()} key={cell.id}>
                            {cell.render('Cell')}
                          </td>
                        ))}
                      </tr>
                    )}
                  </React.Fragment>
                );
              case ETableTypes.securityConcepts:
                return (
                  <React.Fragment key={row.id}>
                    {row.values.subRequirement ? (
                      <RequirementDropdown
                        requirement={row.values.subRequirement}
                        showExpandedInfo={row.isExpanded}
                        expander={row.cells[0]}
                      />
                    ) : (
                      <tr {...row.getRowProps()} className="table-row" key={row.id}>
                        {row.cells.map((cell: CellValue) => (
                          <td className="table-col" {...cell.getCellProps()} key={cell.id}>
                            {cell.render('Cell')}
                          </td>
                        ))}
                      </tr>
                    )}
                  </React.Fragment>
                );
              case ETableTypes.threats:
                return (
                  <React.Fragment key={row.id}>
                    <tr {...row.getRowProps()} className="table-row">
                      {row.cells.map((cell: CellValue) => (
                        <td className="table-col" {...cell.getCellProps()} key={cell.column.id}>
                          {cell.render('Cell')}
                        </td>
                      ))}
                    </tr>
                    {row.isExpanded && (
                      <tr>
                        <td colSpan={99}>
                          <PropertiesCard type={EThreatsControlsPath.threats} content={row.values.details} />
                        </td>
                      </tr>
                    )}
                  </React.Fragment>
                );

              case ETableTypes.controlGroups:
                return (
                  <React.Fragment key={row.id}>
                    {row.values.control ? (
                      <ControlListDropdown
                        control={row.values.control}
                        showExpandedInfo={row.isExpanded}
                        expander={row.cells[0]}
                      />
                    ) : (
                      <tr key={row.id} {...row.getRowProps()} className="table-row">
                        {row.cells.map((cell: CellValue) => (
                          <td className="table-col controls" {...cell.getCellProps()} key={cell.column.id}>
                            {cell.render('Cell')}
                          </td>
                        ))}
                      </tr>
                    )}
                  </React.Fragment>
                );

              case ETableTypes.feeds:
                return (
                  <React.Fragment key={row.id}>
                    <tr key={row.id} {...row.getRowProps()} className="table-row">
                      {row.cells.map((cell: CellValue) => (
                        <td className="table-col" {...cell.getCellProps()} key={cell.column.id}>
                          {cell.render('Cell')}
                        </td>
                      ))}
                    </tr>
                    {row.isExpanded && (
                      <tr>
                        <td colSpan={99}>
                          <FeedsECUDropdown ECUs={row.values.ECUs} />
                        </td>
                      </tr>
                    )}
                  </React.Fragment>
                );

              default:
                return (
                  <tr key={row.id} {...row.getRowProps()} className="table-row">
                    {row.cells.map((cell: CellValue) => (
                      <td key={cell.id} className="table-col" {...cell.getCellProps()}>
                        {cell.render('Cell')}
                      </td>
                    ))}
                  </tr>
                );
            }
          })}
        </tbody>
      </table>
    </>
  );
};
