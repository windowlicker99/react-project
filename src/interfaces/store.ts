import { ETableTypes, IColumn, ICurrentElement, IElementsTabsColumnsList, IUser } from '@interfaces/interfaces';

export interface IAuth {
  user: IUser;
}
export interface IPagination {
  nextPage: () => void;
  prevPage: () => void;
  pickPage: (indexPage: number) => void;
  disablePrev: boolean;
  disableNext: boolean;
  currentPage: number;
  pageAmount: number;
  currentTab: string;
  totalElement: number;
  totalElementOnPage: number;
  showedElements: number;
}

export interface IVehicleStore {
  currentVehicle: string;
}

export interface IHeaderStore {
  subtitles: string[];
}

export interface IChangeTypeAction {
  type: ETableTypes;
}

export interface IUpdateDisplayedColumnsAction {
  displayColumnsList: IDisplayColumnsList;
}

export interface ICustomizeColumnState {
  type: ETableTypes;
  columns: IElementsTabsColumnsList;
}

export interface IStoreState {
  auth: IAuth;
  pagination: IPagination;
  vehicles: IVehicleStore;
  customizeColumns: ICustomizeColumnState;
  header: IHeaderStore;
  element: IElementStore;
}

export interface IDisplayColumnsList {
  [key: string]: boolean;
}

export interface ICustomizeColumnAction {
  type: ETableTypes;
  displayColumnsList: IDisplayColumnsList;
}

export interface ICustomizeColumnsState {
  [key: string]: IColumn[];
}

export interface IElementStore {
  currentElement: ICurrentElement;
}
