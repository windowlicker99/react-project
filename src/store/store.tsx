import { combineReducers } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage';
import { persistStore, persistReducer } from 'redux-persist';
import { authReducer } from '@store/slices/authSlice';
import { columnsReducer } from '@store/slices/columnsSlice';
import { paginationReducer } from '@store/slices/paginationSlice';
import { vehiclesReducer } from '@store/slices/vehicleSlice';
import { headerReducer } from '@store/slices/headerSlice';
import { elementReducer } from '@store/slices/elementSlice';

const reducers = combineReducers({
  auth: authReducer,
  vehicles: vehiclesReducer,
  customizeColumns: columnsReducer,
  pagination: paginationReducer,
  header: headerReducer,
  element: elementReducer,
});

const persistConfig = {
  key: 'root',
  storage,
  whiteList: ['auth', 'customizeColumns'],
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) => [
    ...getDefaultMiddleware({
      serializableCheck: false,
    }),
  ],
});

export const persistor = persistStore(store);
