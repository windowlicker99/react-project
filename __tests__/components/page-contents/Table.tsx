/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Table } from '@components/page-contents/Table';
import { ETableTypes } from '@interfaces/interfaces';
import { store } from '@store/store';

describe('Table tests', () => {
  const columns = [
    {
      Header: 'test',
      accessor: 'test',
    },
  ];

  it('Should render', () => {
    const { container } = render(
      <Provider store={store}>
        <Table changeLoadingState={jest.fn()} columns={columns} tableType={ETableTypes.domains} showHeader />
      </Provider>
    );
    expect(container.getElementsByClassName('table').length).toEqual(1);
  });

  it('Should match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Table changeLoadingState={jest.fn()} columns={columns} tableType={ETableTypes.domains} showHeader />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
