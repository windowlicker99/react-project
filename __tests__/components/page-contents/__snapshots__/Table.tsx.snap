// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table tests Should match snapshot 1`] = `
<DocumentFragment>
  <table
    class="table"
    role="table"
  >
    <thead
      class="table-header"
    >
      <tr
        class="table-row"
        role="row"
      >
        <th
          class="table-col"
          colspan="1"
          role="columnheader"
        >
          <div
            class="table-header-inner"
          >
            <span
              class="table-header-title"
            >
              test
            </span>
            <div
              class="table-filter"
            />
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      class="table-body"
      role="rowgroup"
    />
  </table>
</DocumentFragment>
`;
